File: app/src/main/java/com/gb/canibuyit/util/DialogUtils.java
package com.gb.canibuyit.util;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;

public class DialogUtils {

    /**
     * @param onSave   will be run if user selects Yes
     * @param onFinish will be run if user selects Yes and onSave returns true OR if
     *                 user selects Discard (in which case onSave is ignored)
     */
    public static AlertDialog getSaveOrDiscardDialog(Context context, final String message, final Executable onSave, final Runnable onFinish) {
        return getDialog(context, "Save changes?", message,"Save",
                new ClickRunner.TaskBuilder(onSave).onSuccess(onFinish).build(), "Discard",
                new ClickRunner.Builder().setOnDiscard(onFinish).build(), android.R.string.cancel, null);
    }


    public static AlertDialog getErrorDialog(Context context, String message) {
        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        builder.setTitle("Error").setMessage(message).setPositiveButton(android.R.string.ok, null);
        return builder.create();
    }

    /**
     * @param title                   CharSequence or resource id
     * @param positiveButton          CharSequence or resource id
     * @param onPositiveButtonClicked
     * @param negativeButton          CharSequence or resource id
     * @param onNegativeButtonClicked
     * @param fuckOffButton           CharSequence or resource id
     * @param onFuckOffButtonClicked
     */
    public static AlertDialog getDialog(Context context, Object title, Object message, Object positiveButton, ClickRunner onPositiveButtonClicked,
                                        Object negativeButton, ClickRunner onNegativeButtonClicked, Object fuckOffButton, ClickRunner onFuckOffButtonClicked) {
        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        if (title != null) {
            if (title instanceof CharSequence) {
                builder.setTitle((CharSequence) title);
            } else if (title instanceof Integer) {
                builder.setTitle((int) title);
            } else {
                throw new IllegalArgumentException("Wrong title type in DialogUtils.getDialog(...)");
            }
        }
        if (message != null) {
            if (message instanceof CharSequence) {
                builder.setMessage((CharSequence) message);
            } else if (title instanceof Integer) {
                builder.setMessage((int) message);
            } else {
                throw new IllegalArgumentException("Wrong title type in DialogUtils.getDialog(...)");
            }
        }
        if (positiveButton != null) {
            if (positiveButton instanceof CharSequence) {
                builder.setPositiveButton((CharSequence) positiveButton, onPositiveButtonClicked);
            } else if (positiveButton instanceof Integer) {
                builder.setPositiveButton((int) positiveButton, onPositiveButtonClicked);
            } else {
                throw new IllegalArgumentException("Wrong positive button text type in DialogUtils.getDialog(...)");
            }
        }
        if (negativeButton != null) {
            if (negativeButton instanceof CharSequence) {
                builder.setNegativeButton((CharSequence) negativeButton, onNegativeButtonClicked);
            } else if (negativeButton instanceof Integer) {
                builder.setNegativeButton((int) negativeButton, onNegativeButtonClicked);
            } else {
                throw new IllegalArgumentException("Wrong negative button text type in DialogUtils.getDialog(...)");
            }
        }
        if (fuckOffButton != null) {
            if (fuckOffButton instanceof CharSequence) {
                builder.setNeutralButton((CharSequence) fuckOffButton, onFuckOffButtonClicked);
            } else if (fuckOffButton instanceof Integer) {
                builder.setNeutralButton((int) fuckOffButton, onFuckOffButtonClicked);
            } else {
                throw new IllegalArgumentException("Wrong fuck off button text type in DialogUtils.getDialog(...)");
            }
        }
        return builder.create();
    }

    public static abstract class Executable {
        public abstract boolean run();
    }

    private static class ClickRunner implements DialogInterface.OnClickListener {

        private final Executable conditionalTask;
        private final Runnable[] onSuccess;
        private final Runnable[] onFail;

        ClickRunner(Builder builder) {
            conditionalTask = null;
            onSuccess = builder.onDiscard;
            onFail = null;
        }

        ClickRunner(TaskBuilder taskBuilder) {
            conditionalTask = taskBuilder.conditionalTask;
            onSuccess = taskBuilder.onSuccess;
            onFail = taskBuilder.onFail;
        }

        @Override
        public void onClick(DialogInterface dialog, int which) {
            boolean result = conditionalTask == null || conditionalTask.run();

            if (result && onSuccess != null) {
                for (Runnable r : onSuccess) {
                    try {
                        r.run();
                    } catch (Throwable t) {
                        Logger.INSTANCE.d(ClickRunner.class.getName(), t);
                    }
                }
            } else if (onFail != null) {
                for (Runnable r : onFail) {
                    try {
                        r.run();
                    } catch (Throwable t) {
                        Logger.INSTANCE.d(ClickRunner.class.getName(), t);
                    }
                }
            }
        }

        static class Builder {
            private Runnable[] onDiscard;

            Builder setOnDiscard(Runnable... onDiscard) {
                this.onDiscard = onDiscard;
                return this;
            }

            ClickRunner build() {
                return new ClickRunner(this);
            }
        }

        static class TaskBuilder {
            private Executable conditionalTask;
            private Runnable[] onSuccess;
            private Runnable[] onFail;

            TaskBuilder(Executable onClickTask) {
                conditionalTask = onClickTask;
            }

            TaskBuilder onSuccess(Runnable... onSuccess) {
                this.onSuccess = onSuccess;
                return this;
            }

            public TaskBuilder onFail(Runnable... onFail) {
                this.onFail = onFail;
                return this;
            }

            ClickRunner build() {
                return new ClickRunner(this);
            }
        }
    }
}

File: app/src/main/java/com/gb/canibuyit/util/DialogUtils.kt
package com.gb.canibuyit.util

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface

object DialogUtils {

    /**
     * @param negativeTask will be run if user taps the Discard button
     */
    fun getDiscardDialog(context: Context, message: String, negativeTask: () -> Unit): AlertDialog {
        return getDialog(
                context,
                title = "Cannot save changes",
                message = message,
                negativeButton = "Discard",
                onNegativeButtonClicked = ClickRunner(negativeTask),
                neutralButton = android.R.string.cancel,
                onNeutralButtonClicked = null)
    }

    /**
     * @param positiveTask   will be run if user taps Save
     */
    fun getSaveDialog(context: Context, message: String?, positiveTask: () -> Unit): AlertDialog {
        return getDialog(
                context,
                title = "Save changes?",
                message = message,
                positiveButton = "Save",
                onPositiveButtonClicked = ClickRunner(positiveTask),
                neutralButton = android.R.string.cancel,
                onNeutralButtonClicked = null)
    }

    /**
     * @param positiveConditionalTask   will be run if user taps Save
     * @param onFinish will be run if user selects Save and positiveConditionalTask returns true OR if
     * user selects Discard (in which case positiveConditionalTask is ignored)
     */
    fun getSaveOrDiscardDialog(context: Context, message: String?,
                               positiveConditionalTask: () -> Boolean,
                               onFinish: () -> Unit): AlertDialog {
        return getDialog(
                context,
                title = "Save changes?",
                message = message,
                positiveButton = "Save",
                onPositiveButtonClicked = ClickRunner(onFinish, positiveConditionalTask),
                negativeButton = "Discard",
                onNegativeButtonClicked = ClickRunner(onFinish),
                neutralButton = android.R.string.cancel,
                onNeutralButtonClicked = null)
    }

    /**
     * @param title                   CharSequence or resource id
     * @param positiveButton          CharSequence or resource id
     * @param onPositiveButtonClicked
     * @param negativeButton          CharSequence or resource id
     * @param onNegativeButtonClicked
     * @param neutralButton           CharSequence or resource id
     * @param onNeutralButtonClicked
     */
    private fun getDialog(context: Context, title: Any?, message: Any?, positiveButton: Any? = null,
                  onPositiveButtonClicked: ClickRunner? = null,
                  negativeButton: Any? = null, onNegativeButtonClicked: ClickRunner? = null,
                  neutralButton: Any?, onNeutralButtonClicked: ClickRunner?): AlertDialog {
        val builder = AlertDialog.Builder(context)
        if (title != null) {
            when (title) {
                is CharSequence -> builder.setTitle(title as CharSequence?)
                is Int -> builder.setTitle(title)
                else -> throw IllegalArgumentException("Wrong title type in DialogUtils.getDialog(...)")
            }
        }
        if (message != null) {
            when {
                message is CharSequence -> builder.setMessage(message as CharSequence?)
                title is Int -> builder.setMessage(message as Int)
                else -> throw IllegalArgumentException("Wrong title type in DialogUtils.getDialog(...)")
            }
        }
        if (positiveButton != null) {
            when (positiveButton) {
                is CharSequence -> builder.setPositiveButton(positiveButton as CharSequence?, onPositiveButtonClicked)
                is Int -> builder.setPositiveButton(positiveButton, onPositiveButtonClicked)
                else -> throw IllegalArgumentException(
                        "Wrong positive button text type in DialogUtils.getDialog(...)")
            }
        }
        if (negativeButton != null) {
            when (negativeButton) {
                is CharSequence -> builder.setNegativeButton(negativeButton as CharSequence?, onNegativeButtonClicked)
                is Int -> builder.setNegativeButton(negativeButton, onNegativeButtonClicked)
                else -> throw IllegalArgumentException(
                        "Wrong negative button text type in DialogUtils.getDialog(...)")
            }
        }
        if (neutralButton != null) {
            when (neutralButton) {
                is CharSequence -> builder.setNeutralButton(neutralButton as CharSequence?, onNeutralButtonClicked)
                is Int -> builder.setNeutralButton(neutralButton, onNeutralButtonClicked)
                else -> throw IllegalArgumentException(
                        "Wrong fuck off button text type in DialogUtils.getDialog(...)")
            }
        }
        return builder.create()
    }

    class ClickRunner @JvmOverloads constructor(private var onSuccess: (() -> Unit)? = null,
                                                private var conditionalTask: (() -> Boolean)? = null,
                                                private var onFail: (() -> Unit)? = null) :
            DialogInterface.OnClickListener {

        override fun onClick(dialog: DialogInterface, which: Int) {
            conditionalTask?.let {
                if (it.invoke()) onSuccess?.invoke() else onFail?.invoke()
            } ?: onSuccess?.invoke()
        }
    }
}
